File: examples/03_navigation.cpp
// 03 - Navigating the half-edge structure (C++).
//   The navigation primitives, the circulators, and the lazy ranges.
#include "simplemesh/Mesh.h"
#include <iostream>

int main() {
    sm::Mesh m;
    auto v0 = m.add_vertex({0, 0, 0});
    auto v1 = m.add_vertex({1, 0, 0});
    auto v2 = m.add_vertex({1, 1, 0});
    auto v3 = m.add_vertex({0, 1, 0});
    m.add_triangle(v0, v1, v2);
    m.add_triangle(v0, v2, v3);

    auto h = m.find_halfedge(v0, v2);                 // the shared diagonal
    std::cout << "halfedge v0->v2 : " << h.idx() << "\n";
    std::cout << "  to / from     : " << m.to_vertex(h).idx() << " / " << m.from_vertex(h).idx() << "\n";
    std::cout << "  opposite      : " << m.opposite_halfedge(h).idx() << "\n";
    std::cout << "  next / prev   : " << m.next_halfedge(h).idx() << " / " << m.prev_halfedge(h).idx() << "\n";
    std::cout << "  interior edge : " << (m.is_boundary(m.edge(h)) ? "no" : "yes") << "\n";

    std::cout << "1-ring of v0    :";
    for (auto w : m.vertex_vertices(v0)) std::cout << " " << w.idx();
    std::cout << "\nvalence of v0   : " << m.valence(v0) << "\n";

    // lazy ranges (zero allocation) work in range-for:
    std::cout << "neighbours (lazy):";
    for (auto w : m.vv_range(v0)) std::cout << " " << w.idx();
    std::cout << "\n";
    return 0;
}

File: examples/03_navigation.py
"""03 - Navigating the half-edge structure.

The eight primitives (next/prev/opposite, to/from vertex, face, edge) plus the
circulators and lazy iterators that traverse a neighbourhood.
Run:  PYTHONPATH=build/bin python examples/03_navigation.py
"""
import simplemesh as sm

# two triangles sharing the diagonal v0-v2 (a split quad)
m = sm.Mesh()
v0 = m.add_vertex(sm.Vec3(0, 0, 0))
v1 = m.add_vertex(sm.Vec3(1, 0, 0))
v2 = m.add_vertex(sm.Vec3(1, 1, 0))
v3 = m.add_vertex(sm.Vec3(0, 1, 0))
m.add_triangle(v0, v1, v2)
m.add_triangle(v0, v2, v3)

h = m.find_halfedge(v0, v2)                 # the shared diagonal
print("halfedge v0->v2:", h.idx())
print("  to_vertex      :", m.to_vertex(h).idx())
print("  from_vertex    :", m.from_vertex(h).idx())
print("  opposite       :", m.opposite_halfedge(h).idx())
print("  next / prev    :", m.next_halfedge(h).idx(), "/", m.prev_halfedge(h).idx())
print("  interior edge? :", not m.is_boundary_edge(m.edge(h)))

print("1-ring of v0     :", [w.idx() for w in m.vertex_vertices(v0)])
print("faces around v0  :", [f.idx() for f in m.vertex_faces(v0)])
print("valence of v0    :", m.valence(v0))

# lazy iterators (no allocation) work in for-loops:
print("all vertices     :", [w.idx() for w in m.vertices()])
print("neighbours of v0 :", [w.idx() for w in m.vv(v0)])
